### ISSUE TITLE
Maf Data conversion from 10 to 12 bit analog data on the ESP32

### ISSUE THREAD

**Issue by DeeEmm** [DeeEmm/DIY-Flow-Bench#79]
Old analog conversion on Mega / Uno was 10 bit resolution (1024 data points) but ESP32 is capable of 12 bit resolution (4096 data points). In the short term this means that the old code does not work. In the long term it may mean updating both the code and the mafData files to take advantage of the better data resolution.

Proposal is to get code working with existing mafData but with a view to be able to utilise higher resolution data files should they be available.

**Comment by DeeEmm** [comment]
Need to detect size of mafData array and then do some maths to interpolate data values. 

For a 12 bit array we can just use the raw value, but for a 10 bit array our 12 bit reference will not work, so we need to work out which two values fall either side of our reference value and then do some interpolation to calculate the actual value. We could of course limit the reference value to 10 bits (probably a good idea initially) but then we lose that extra resolution.

We are already doing a similar thing with the key>value style mafData array as this is not a fixed length array.

There we are using the following interpolation formula:

```
Y=Y0+(((X-X0)(Y1-Y0))/(X1-X0)) 

where Y = flow and X = Volts
```

We might even be able to recycle this code and move it into a function

**Comment by DeeEmm** [comment]
Also need to convert from 5v to 3.3V.

Old key>value array data range is from 0-5000 (0-5v) which means that we can no longer use raw analog values to directly interrogate the data array. We first need to scale the reference value, the data or potentially both.

So it looks like the best course of action is going to be to to re-write the MAF data file handling. This will also be in line with Issue #76 which requires a full review and update of the sensor code.

So It is going to be better to use an arbitrary array length that also has an arbitrary scaling. This way we do not care how many data points we have, or what voltage they were recorded at. or even if they relate to voltage. By doing this, we should also then be able to accommodate the frequency based MAF data that we need to integrate as part of issue #26.

I think that we should firstly move to a 2D array for all data. i.e.

```
{{key1, val1} , {key2,val2}, {etc..}}
```

It will be easy enough to transpose the existing array data that we have for analog sensors. where the key value simply takes the existing array reference value.

```
{1,2,5,7,9}

becomes...

{
  {0,1},
  {1,2},
  {2,5},
  {3,7},
  {4,9}
}
```

This move commonalises all Maf Data arrays into the same basic structure which makes the code easier.

We can then use class methods to determine what type of data it is and what scaling it was recorded at. 

I have already converted the mafData files to a class / method style of code structure so that we can have multiple instances of maf data, that all share the same data names and data types without generating pages of compiler errors, so we should end up with something like this...

```
class MafData {

	public:
		MafData();
		String mafSensorType ();
		int MAFdataFormat ();
		int MAFdataUnit ();
		int MAFoutputType();
		long getMafData();
	
		long mafLookupTable[][2];
	
	private:

};
```

We can expand on the methods above with additional methods as required.

**Comment by DeeEmm** [comment]
I've converted the MafData files back to basic header files as I was unable to resolve some clashes with the class definitions. I tried to control this using namespaces but ended digging myself a deeper hole so decide to revert back to the original way of handing this. This has changed the way we interact with the mafData but as all of that stuff is currently in flux changing it is not really an issue.

We still have the array changes noted above and are using a 2D array as a lookup table for all data types (voltage / frequency / etc)

How the array is processed is the same in all instances as it is a simple lookup function.

We take an arbitrary sensor value (millivolts / herts / baked-beans / etc)  and we then look up the corresponding value from the lookup table. This value is in mass-flow (kg/h or g/s). Apart from scaling, the result will always be mass-flow, irrespective of what the input is.

We get the sensor value from the sensors class, so it is there that the analog input value is converted into the relevant format to work with the lookup table. As all mafData variables are exposed to the sensor class it simply uses mafOutputType to determine how to translate the raw sensor value into whatever value the lookup table needs.

In the case of a voltage based sensor I have changed the voltage and input resolution to 3.3v / 12 bit.

```
int mafFlowRaw = analogRead(MAF_PIN);
double mafMillivolts = (mafFlowRaw * (_hardware.getSupplyMillivolts() / 4095.0)) * 1000;
return mafMillivolts;
```

...which concludes this particular issue request. However there is still some work to do to address #26 

/DM
